File: TD/TD4/Code/6AB.cpp
/**
 * 1. Sous-programme -> Calculer la somme des diviseurs et leur produit
 * 2. Fonction -> Retourner VRAI, si entier est un doublon, FAUX sinon
 */
#include <iostream>

/* Procédure */

void sommeProduit(int n, int &som, int &prod)
{
  som = 0;
  prod = 1;

  for (int i = 1; i <= n; ++i)
    if (n % i == 0) {
      som += i;
      prod *= i;
    }
}

/* Fonction */

bool verifieDoublon(int n)
{
  int som;
  int prod;

  sommeProduit(n, som, prod);

  if (prod % som == 0)
    return true;
  else
    return false;
}

/* Utilisation */

int main()
{
  int n;

  std::cout << "Donne moi une valeur : " << std::flush;
  std::cin >> n;

  if (verifieDoublon(n))
    std::cout << "Oui" << std::endl;
  else
    std::cout << "Non" << std::endl;

  return 0;
}

File: TD/TD4/Code/Python/6AB.py
#!/usr/bin/python
# 1. Calculer la somme des diviseurs et leur produit
# 2. Retourner VRAI, si entier est un doublon, FAUX sinon


def somme_produit(foo):
    somme = 0
    prod = 1
    for i in range(1, foo + 1):
        if foo % i == 0:
            somme += i
            prod *= i
    return somme, prod


def est_doublon(foo):
    somme, prod = somme_produit(foo)
    res = True if prod % somme == 0 else False
    return res


print('Donne moi une valeur : ', end='')
foo = int(input())
print('Oui' if est_doublon(foo) else 'Non')
